Plans data-store: improve unit tests

ciampo opened the pull request:
### Changes proposed in this Pull Request

* Refactor mock data:
  - split into smaller files, divided by data type (e.g. APIs vs Store)
  - removed `MOCK_` prefix
  - slightly changed naming convention
* Add some missing mock data
* Refactored existing tests, mainly to accommodate for the new mock data naming conventions (and now using the `MockData` namespace for accessing all mock variables)
* Added unit tests for action creators
* Added unit tests for selectors

### Testing instructions

#### Manual tests

- [ ] I left some comments below on the most interesting bits of code
- [ ] Check if there's any missing test case, or if any tests could be testing for the wrong result

#### Automated tests

- [ ] `./node_modules/.bin/jest -c=test/packages/jest.config.js packages/data-stores/src/plans` — all tests passing
- [ ] `./node_modules/.bin/jest -c=test/packages/jest.config.js --coverage --collectCoverageFrom='["src/plans/*.ts", "!src/plans/{index,types}.ts"]' packages/data-stores/src/plans` — 100% coverage on all files


### Screenshots

### Before

<img width="876" alt="Screenshot 2021-02-26 at 16 11 54" src="https://user-images.githubusercontent.com/1083581/109317819-69eeba80-784d-11eb-986d-9e3bf0664e47.png">

### After

<img width="879" alt="Screenshot 2021-02-26 at 16 09 51" src="https://user-images.githubusercontent.com/1083581/109317511-11b7b880-784d-11eb-8b7d-7ee968db13d2.png">

Closes #49920

ciampo committed:
Add action generators unit tests for the plans data-store

ciampo committed:
Remove unnecessary cast to "any"

ciampo committed:
Refactor: split mock data into files, add missing data, add utils, use * imports

ciampo committed:
Mock wpcom-request-controls import directly

ciampo committed:
Revert previous commit (that mocked module was actually an internal import)

ciampo committed:
Shorter type declaration

ciampo committed:
Plans data store: selectors unit tests for selectors that don't involve resolvers

ciampo committed:
Move jest.mocks after mock data declaration

ciampo committed:
Rename all 'test*' variables to 'mock*'

ciampo committed:
Mock @wordpress/data module, to avoid calling resolvers when testing selectors

ciampo committed:
Add missing selectors unit tests

ciampo commented on packages/data-stores/src/plans/index.ts:
```diff
@@ -50,7 +50,7 @@ export function register(): typeof STORE_KEY {
 			actions,
 			controls: controls as any,
-			reducer: reducer as any,
+			reducer,
```
This wasn't needed

ciampo commented on packages/data-stores/src/plans/selectors.ts:
```diff
@@ -14,7 +15,6 @@ import {
 	STORE_KEY,
 	FREE_PLAN_PRODUCT_ID,
 } from './constants';
-import deprecate from '@wordpress/deprecated';
```
Moved to external dependencies (see above)

StefanNieuwenhuis replied:
Nitpick mode: Scope creep?

ciampo replied:
Reverted in https://github.com/Automattic/wp-calypso/pull/50492/commits/6a44b523402b69d8986278e0eb065797e2486481, and opened #50666 to include these changes in a separate PR.

ciampo commented on packages/data-stores/src/plans/selectors.ts:
```diff
@@ -131,26 +132,28 @@ export const getPlanByPath = (
 
 	return select( STORE_KEY )
 		.getSupportedPlans( locale )
-		.find( ( plan ) => plan.periodAgnosticSlug === planProduct?.periodAgnosticSlug );
+		.find( ( plan ) => plan.periodAgnosticSlug === planProduct.periodAgnosticSlug );
```
The `?.` check wasn't necessary, because of the typeguard check a few lines above

StefanNieuwenhuis replied:
Nitpick mode: Scope creep?

ciampo replied:
Reverted in https://github.com/Automattic/wp-calypso/pull/50492/commits/6fe2a1e0f32d57bdbee7ec12eaeed06e5489b1a2, and opened #50666 to include these changes in a separate PR.

ciampo commented on packages/data-stores/src/plans/selectors.ts:
```diff
@@ -131,26 +132,28 @@ export const getPlanByPath = (
-	);
-	return planProduct;
+
+	return select( STORE_KEY )
+		.getPlansProducts()
+		.find( ( product ) => {
+			const matchesSlug = product.periodAgnosticSlug === periodAgnosticSlug;
+			// The billing period doesn't matter when dealing with free plan
+			const matchesBillingPeriod =
+				periodAgnosticSlug === TIMELESS_PLAN_FREE || product.billingPeriod === billingPeriod;
+
+			return matchesSlug && matchesBillingPeriod;
+		} );
```
This is just a refactor of the selector function, without changing its logic

StefanNieuwenhuis replied:
Although the selector's logic hasn't changed, I'm not very comfortable with this change since I believe this needs to be addressed in a separate and dedicated PR. Can you please revert this change and fix it in another PR?

ciampo replied:
Reverted in https://github.com/Automattic/wp-calypso/pull/50492/commits/6fe2a1e0f32d57bdbee7ec12eaeed06e5489b1a2, and opened #50666 to include these changes in a separate PR.

ciampo commented on packages/data-stores/src/plans/test/resolvers.ts:
```diff
@@ -30,6 +13,8 @@ jest.mock( 'wpcom-proxy-request', () => ( {
 	__esModule: true,
 } ) );
```
I decided not to mock the `wpcom-request-controls` module (see [commit](https://github.com/Automattic/wp-calypso/pull/50492/commits/0101d833745fa872724e458401b66b992c52349d)), since its functions are still part of the `data-stores` package.

I know that by not mocking these functions, these unit tests become also partially integration tests — but I think it's fine in this case.

StefanNieuwenhuis replied:
We probably have to update our resolvers somewhere in the future (https://github.com/Automattic/wp-calypso/pull/40371#issuecomment-787843426) since `wpcom-request-controls` breaks when we migrate to OAuth.

ciampo replied:
That's correct, but I would wait for directions from the Calypso folks before making any changes to that part of the `data-stores` package

ciampo commented on packages/data-stores/src/plans/test/selectors.ts:
```diff
@@ -0,0 +1,326 @@
+};
+
+// Mocks
+jest.mock( '@wordpress/deprecated', () => {
+	return jest.fn();
+} );
```
Mocking `'@wordpress/deprecated'` because:
- it avoids polluting the console output with console.warnings
- we can spy on it, and make sure it's called the right number of times / with the correct arguments, to prove the fact that deprecated functions are still going to be generating a warning

ciampo commented on packages/data-stores/src/plans/test/selectors.ts:
```diff
@@ -0,0 +1,326 @@
+jest.mock( '@wordpress/deprecated', () => {
+	return jest.fn();
+} );
+jest.mock( '@wordpress/data', () => ( {
+	// Mocking `select` allows unit tests to run in isolation for selectors,
+	// without the corresponding resolvers being called
+	select: () => ( {
+		getPlansProducts: () => mockPlanProducts,
+		getSupportedPlans: ( locale ) => mockPlans[ locale ] ?? [],
+	} ),
+} ) );
```
Mocking the `select` function from `@wordpress/data` is essential for testing the selectors in isolation, since it prevents the corresponding resolvers from being called during unit testing

ciampo commented on packages/data-stores/src/plans/test/selectors.ts:
```diff
@@ -0,0 +1,326 @@
+		expect( Selectors.getPrices( mockState, TEST_LOCALE_1 ) ).toEqual( expectedPrices );
+		expect( Selectors.getPrices( mockState, TEST_LOCALE_2 ) ).toEqual( expectedPrices );
+
+		// Make sure function is correctly flagged as deprecated
+		const expectedCallArguments = [
+			'getPrices',
+			{
+				alternative: 'getPlanProduct().price',
+			},
+		];
+		expect( deprecate ).toHaveBeenCalledTimes( 2 );
+		expect( deprecate ).toHaveBeenNthCalledWith( 1, ...expectedCallArguments );
+		expect( deprecate ).toHaveBeenNthCalledWith( 2, ...expectedCallArguments );
```
See https://github.com/Automattic/wp-calypso/pull/50492#discussion_r583702273

StefanNieuwenhuis commented on packages/data-stores/src/plans/index.ts:
```diff
@@ -50,7 +50,7 @@ export function register(): typeof STORE_KEY {
 			resolvers,
 			actions,
 			controls: controls as any,
```
If the reducer doesn't have to be typed as `any` anymore, why do we need it for `controls`?

ciampo replied:
I believe that the `reducer` type definition initially were causing errors, but they probably improved over time, and the cast to `any` was not necessary.

Instead, removing the `any` type from `controls` causes this TypeScript error:

```
Type '{ readonly WPCOM_REQUEST: ({ request }: ReturnType<typeof wpcomRequest>) => Promise<unknown>; readonly FETCH_AND_PARSE: ({ resource, options }: ReturnType<typeof fetchAndParse>) => Promise<{ ok: boolean; body: any; }>; readonly RELOAD_PROXY: () => void; readonly REQUEST_ALL_BLOGS_ACCESS: () => Promise<...>; readonly...' is not assignable to type '{ [k: string]: (action: AnyAction) => any; }'.
  Property 'WPCOM_REQUEST' is incompatible with index signature.
    Type '({ request }: ReturnType<typeof wpcomRequest>) => Promise<unknown>' is not assignable to type '(action: AnyAction) => any'.
      Types of parameters '__0' and 'action' are incompatible.
        Property 'request' is missing in type 'AnyAction' but required in type '{ readonly type: "WPCOM_REQUEST"; readonly request: WpcomRequestParams; }'.ts(2322)
index.ts(12, 29): 'request' is declared here.
index.d.ts(66, 5): The expected type comes from property 'controls' which is declared here on type 'StoreConfig<CombinedState<{ features: Record<string, Record<string, PlanFeature>>; featuresByType: Record<string, FeaturesByType[]>; planProducts: PlanProduct[]; plans: Record<...>; }>>'
```

I am not very knowledgeable about `@wordpress/data` types — plus typing `controls` to `any` seems to be a common practice in other data-stores too.

I would leave this line as is for now.

StefanNieuwenhuis commented on packages/data-stores/src/plans/test/actions.ts:
```diff
@@ -0,0 +1,64 @@
+import * as MockData from '../mock';
+import { buildPlanFeaturesDict } from '../test-utils';
+
+const TEST_LOCALE = 'test-locale';
```
Bit of nitpicking but for the sake of consistency, can you change its name to something that includes `mock`? I'm suggesting this because we're using this "_prefix_" for all mocked data in our tests.

ciampo replied:
Done in https://github.com/Automattic/wp-calypso/pull/50492/commits/e8212f4d122a4fb2a8de244541933aa52d482582 and https://github.com/Automattic/wp-calypso/pull/50492/commits/8ce121bb0651311685bbb6bbd89465b51631f813

StefanNieuwenhuis commented on packages/data-stores/src/plans/test/reducer.ts:
```diff
@@ -3,21 +3,10 @@
+import { buildPlanFeaturesDict } from '../test-utils';
 
-const LOCALE_EN = 'en';
+const TEST_LOCALE = 'test-locale';
```
Same here (see my previous comment).

ciampo replied:
Done in https://github.com/Automattic/wp-calypso/pull/50492/commits/e8212f4d122a4fb2a8de244541933aa52d482582 and https://github.com/Automattic/wp-calypso/pull/50492/commits/8ce121bb0651311685bbb6bbd89465b51631f813

StefanNieuwenhuis commented on packages/data-stores/src/plans/test/resolvers.ts:
```diff
@@ -30,6 +13,8 @@ jest.mock( 'wpcom-proxy-request', () => ( {
+// const TEST_LOCALE = 'test-locale';
+
 describe( 'getSupportedPlans', () => {
 	it( 'calls setFeatures, setFeaturesByType, and setPlans after fetching plans', () => {
```
According to strict Unit Testing rules, these should be 3 separate tests. What do you think?

ciampo replied:
I see your point, but I'm not sure how to split this test, since it relies on testing the resolver as a generator function, calling `next()` on each.

For now, I've wrapped this test into one more resolvers `describe()` call, and renames the test to a better name in https://github.com/Automattic/wp-calypso/pull/50492/commits/6a2c1cb1af6a86e1cd6f11c8fe520b415f02aef2

StefanNieuwenhuis commented on packages/data-stores/src/plans/test/resolvers.ts:
```diff
@@ -30,6 +13,8 @@ jest.mock( 'wpcom-proxy-request', () => ( {
 	__esModule: true,
 } ) );
 
+// const TEST_LOCALE = 'test-locale';
```
Can you please remove this?

```suggestion
```

ciampo replied:
Good spot! I was actually meaning to use that variable — I uncommented it and used it in https://github.com/Automattic/wp-calypso/pull/50492/commits/565f71dbe20297cf9be7a08a2bb6ac22167eca37

StefanNieuwenhuis commented on packages/data-stores/src/plans/test/resolvers.ts:
```diff
@@ -66,56 +51,43 @@ describe( 'getSupportedPlans', () => {
 		} );
 
 		expect( iter.next().value ).toEqual( {
 			locale: 'en',
```
Why don't we use a mocked value here?

ciampo replied:
Good question — I refactored the code to use a mocked value in https://github.com/Automattic/wp-calypso/pull/50492/commits/565f71dbe20297cf9be7a08a2bb6ac22167eca37

StefanNieuwenhuis commented on packages/data-stores/src/plans/test/selectors.ts:
```diff
@@ -0,0 +1,326 @@
+import { TIMELESS_PLAN_ECOMMERCE, TIMELESS_PLAN_FREE, FREE_PLAN_PRODUCT_ID } from '../constants';
+import type { State } from '../reducer';
+
+// Test data
```
I think it makes more sense to move all the mocks to the `/mock/` folder too? What do you think?

ciampo replied:
The `mock/` folder contains sample data, while the `mock` variables at the top of this file are simulating different combinations of feature/plans and locales.

They are just recreating the reducer `state` by using mock data, similarly to what we [already do in many other tests](https://github.com/Automattic/wp-calypso/pull/50492/files#diff-705af06b7e1aeb4a5996de8d769e24dad67a1d205e160026e4bcba8b0faf8e28R26-R35). I almost see them as a test implementation detail, not mock data.

StefanNieuwenhuis commented on packages/data-stores/src/plans/test/selectors.ts:
```diff
@@ -0,0 +1,326 @@
+
+// Tests
+describe( 'Plans selectors', () => {
+	it( 'getFeatures', () => {
```
Same as above...According to strict Unit Testing rules, these (and all others in this file) should all be separate tests. I'm not sure if I find it necessary here and I'm curious about your opinion. What do you think?

StefanNieuwenhuis commented on packages/data-stores/src/plans/test/selectors.ts:
```diff
@@ -0,0 +1,326 @@
+				alternative: 'getPlanProduct().price',
+			},
+		];
+		expect( deprecate ).toHaveBeenCalledTimes( 2 );
+		expect( deprecate ).toHaveBeenNthCalledWith( 1, ...expectedCallArguments );
+		expect( deprecate ).toHaveBeenNthCalledWith( 2, ...expectedCallArguments );
```
What value do these test cases add? In other words: Are we testing our selectors or `@wordpress/deprecated `?

ciampo replied:
We're testing that this specific selector is correctly flagged as deprecated by checking that the function responsible for the warning is being called correctly. We're not testing `@wordpress/deprecated` at all, since [it's mocked](https://github.com/Automattic/wp-calypso/pull/50492#discussion_r583702273).

StefanNieuwenhuis commented on packages/data-stores/src/plans/test-utils/index.ts:
```diff
@@ -0,0 +1,15 @@
+ */
+import type { PlanFeature } from '../types';
+
+export const buildPlanFeaturesDict = (
```
❤️  this method and I hope that we can extend our testing utils in the future to make generating mocks easier!

StefanNieuwenhuis reviewed (dismissed):
I've noticed the following:
* `it(...)` & `test(...)` are used both. To be consistent, we should stick with one. 

### On our testing _design_ language
I have a few questions about our testing _design_ language.

1. How do we use `describe()` -- how much nesting do we want to add?
2. Do we want to follow [`BDD` (language) concepts](https://softwareengineering.stackexchange.com/a/152233) (i.e. prefix every test case w/ `should`)?

ciampo committed:
Rename "TEST_LOCALE" to "MOCK_LOCALE"

ciampo committed:
Use mock locale value for resolver tests

ciampo committed:
Rename "TEST_LOCALE_1" and "TEST_LOCALE_2" to "MOCK_LOCALE_1" and "MOCK_LOCALE_2"

ciampo committed:
Refactor resolvers suite/tests structure and names

ciampo committed:
Action creators:

- wrap each action creator into its own describe
- rename "test" to "it"
- write test name following BDD concepts (i.e. "should ...")

ciampo committed:
Reducer:

- rename inner "describe"s to the reducer slice variable
- rename tests to follow BDD concepts ("should...")

ciampo committed:
Selectors:
- wrap each selector into its own describe
- write test names following BDD concepts (i.e. "should ...")
- reset mocks before each test

ciampo commented:
Hey @StefanNieuwenhuis , could you have another look at this PR?


I've addressed all of the feedback that you left in the inline comments, and here are my answers to what you wrote in the main review comment:

> `it(...)` & `test(...)` are used both. To be consistent, we should stick with one.

Good point. I renamed our tests to always use `it()`

> How do we use `describe()` -- how much nesting do we want to add?

I decided to use 2 layers of `describe()` nesting:

1. The outer `describe()` declares what portion of the store is being tested (e.g. action creators, reducer, resolver, selectors)
2. The inner `describe()` declares which particular selector/action creator/resolver/reducer is being tested
3. All tests are nested under 2 `describe()` calls

<details>
  <summary>See screenshot of the new `describe()` nesting structure</summary>
  <img width="779" alt="Screenshot 2021-03-02 at 21 41 52" src="https://user-images.githubusercontent.com/1083581/109712271-1d361700-7ba0-11eb-84c4-c6a005e9117f.png">
</details>

> Do we want to follow [`BDD` (language) concepts](https://softwareengineering.stackexchange.com/a/152233) (i.e. prefix every test case w/ `should`)?

I have updated all test names to start with "should..."

alshakero commented on packages/data-stores/src/plans/selectors.ts:
```diff
@@ -161,5 +161,5 @@ export const isPlanFree = ( _: State, planSlug?: PlanSlug ): boolean => {
 };
 
-export const isPlanProductFree = ( _: State, planProductId: number | undefined ): boolean =>
+export const isPlanProductFree = ( _: State, planProductId?: number ): boolean =>
```
This is not the same. Before this change, the function required the param, but the param can be undefined. After this change you could call `isPlanProductFree()` without any params and TS won't complain, which is undesired.

ciampo replied:
That's a great spot — you're totally right about the unwanted side-effect of this change. I reverted it in https://github.com/Automattic/wp-calypso/pull/50492/commits/7f19087c769ff856c1cb65d3aebdbc321f6379e9

alshakero reviewed (changes requested):
Great cleanup and work! I'm requesting one tiny yet deal-breaker change.

---
(the text below is not the reason for requesting changes)

I really appreciate your work to push us in this direction. I'm a little unsure about the methodology though. For example, testing action creators is testing implementation details that are not part of the API. This makes the tests  less resilient to change. This example extends to testing the resolvers, reducers, etc. I realize that action creators' names are part of the API, but what they do alone isn't part of the API.

For data-stores, I think we should aim for testing the exported API and only that. Anything else won't be testing the product exactly, and will give you pain when you modify any of the impl. details without changing the API.

Also, when testing implementation details, the tests lose a lot of value. Because you'll have to modify the tests whenever you make a change, so your PR will have the change, and the change of tests. Meaning you're trusting humans to review both your change and your tests' change, as opposed to using  the tests as a parallel to human reviewers.

Nonetheless, this PR is a good step forward and I think we should merge it.

ciampo committed:
Revert to using explicit `undefined` type for `isPlanProductFree` selector

ciampo commented:
Hey @alshakero , I've addressed the feedback on the requested change.

Regarding the rest of your feedback:

> For data-stores, I think we should aim for testing the exported API and only that. Anything else won't be testing the product exactly, and will give you pain when you modify any of the impl. details without changing the API.

I still believe that there is value in testing the different parts (units) of a data-store in isolation — testing the external APIs would mean to write _integration_ tests of the type "I dispatch an action, and I expect that the value returned by the selector is correct".

Such integration tests wouldn't be easy to write without mocking certain parts of the store (e.g. the way `@wordpress/data` deals with async resolvers and action generators), which would make the tests lose value too. Plus, having more granual unit tests would also help to isolate the source of a breaking change better.

The only execption that I see (and I partially agree with you about) is regarding action creators — we may argue that, especially with TypeScript, those tests are a bit redundant, but I still think that they add a layer of resilience against regressions. An alternative way of testing for action creators would be to use snapshots instead — basically making sure that any change to action creators needs to be explicitly accompanied by a change in the snapshots.

> Also, when testing implementation details, the tests lose a lot of value.

We're testing each resolver, selector, reducer and action creator in isolation, and we're making sure that each of these functions are functioning as expected — I don't think that we're testing their implementation details

> Because you'll have to modify the tests whenever you make a change, so your PR will have the change, and the change of tests. Meaning you're trusting humans to review both your change and your tests' change, as opposed to using the tests as a parallel to human reviewers.

I think that making changes to the tests (and reviewing those changes) should be part of our duties.

We'd still be using tests as a parallel reviewer — in fact, if a code change breaks a test, it's like having a reviewer rejecting those changes.


I'd love to hear @StefanNieuwenhuis 's (and maybe @tjcafferkey 's) opinions on this subject too!

alshakero commented:
> I still believe that there is value in testing the different parts (units) of a data-store in isolation — testing the external APIs would mean to write integration tests of the type "I dispatch an action, and I expect that the value returned by the selector is correct".

There is no doubt about the value of these tests! But if there is a more valuable alternative (assuming we come to an agreement) we should pursue the alternative. 

> Such integration tests wouldn't be easy to write without mocking certain parts of the store (e.g. the way @wordpress/data deals with async resolvers and action generators), which would make the tests lose value too.

Agree on this one. But in on the long run, I feel we must learn how to do integration testing for the stores and maybe exchange lessons with Gutenberg people.

> Plus, having more granual unit tests would also help to isolate the source of a breaking change better.

While this is true, it comes with a price, if you change the action name from `UPDATE_PLAN` to `MODIFY_PLAN` in both the creator and the reducer, the store is 100% the same. But you'll have to update the tests for virtually no reason.


> We're testing each resolver, selector, reducer and action creator in isolation, and we're making sure that each of these functions are functioning as expected — I don't think that we're testing their implementation details

These functions are implementation details. Aren't they? No one is using the action creators, reducers, nor selectors except the store internals. Reducers are clearly internal, but for selectors and action creators we only expose curried versions, which means the selectors and action creators are implementation details from the store users' perspective.

> I think that making changes to the tests (and reviewing those changes) should be part of our duties.

Definitely, my point is not about duties, it's about the tests being this outsider looking upon your code who judges it. If tests have to change every time your code changes, they're just more code.

My points are largely inspired by [this post](https://kentcdodds.com/blog/making-your-ui-tests-resilient-to-change), I think you'll like it!

alshakero reviewed (approved):
Before:
```sh
Test Suites: 17 passed, 17 total
Tests:       76 passed, 76 total
Snapshots:   0 total
Time:        3.974 s
```

After:
```sh
Test Suites: 19 passed, 19 total
Tests:       123 passed, 123 total // <--- ❤️ 
Snapshots:   0 total
Time:        3.751 s
```

BEAUTY ✨

ciampo commented:
> My points are largely inspired by this post, I think you'll like it!

I love all of Kent's articles!

I think in general we should seek some guidance from core folks about what's the best way to test a `@wordpress/data` store — i feel like there could be lots of learning opportunities

StefanNieuwenhuis commented:
> I'd love to hear @StefanNieuwenhuis 's (and maybe @tjcafferkey 's) opinions on this subject too!

> For data-stores, I think we should aim for testing the exported API and only that. Anything else won't be testing the product exactly, and will give you pain when you modify any of the impl. details without changing the API.

> I still believe that there is value in testing the different parts (units) of a data-store in isolation — testing the external APIs would mean to write integration tests of the type "I dispatch an action, and I expect that the value returned by the selector is correct".

Automated testing is all about confidence. With them in place, we're able to:
* have the confidence that our code works as expected in any environment.
* changes won't affect existing features in unexpected ways.
* new features won't affect existing features in unexpected ways. 

The big question is: How to gain that confidence from automated tests? Which tests should we use?

Kent C. Dodds states "the more your tests resemble the way your software is used, the more confidence they can give you". 

Automate the work a dedicated tester normally would do. A manual tester doesn't care about `classNames`, but he does about the result of a certain action or, in the case of `data-store`, the exported API.

> There is no doubt about the value of these tests! But if there is a more valuable alternative (assuming we come to an agreement) we should pursue the alternative.

If we can have valuable integration tests that:
1. Gain us the same confidence.
2. Are resilient to changes.
3. Are easier to maintain

...then I'm all in for that! Once again, it's all about getting enough confidence from automated tests but I have my doubts if we gain enough confidence by not integration testing resolvers. Resolvers include other API's (e.g. `wpcom-request`) and I believe that we have to test them separately. 

> Also, when testing implementation details, the tests lose a lot of value. Because you'll have to modify the tests whenever you make a change, so your PR will have the change, and the change of tests.

It's still the same with integration tests. Once the API changes, we have to update the tests as well. So this "problem" still persists but I don't think it's necessarily a problem but an opportunity to get a better understanding of the code and the problem it's trying to fix.  While reviewing/running integration tests, you see the code from the end-users POV and, IMHO, this is extremely useful since you have an extra opportunity to improve the API's interface. 

#### Conclusion

* Let's write tests that provide us with enough confidence that our code is resilient to changes and easy to maintain. 
* Let's merge this PR and discuss further steps on testing (e.g. integration vs unit tests).

ciampo committed:
Cover default locale in resolver tests

ciampo commented:
1. Rebased on top of latest `trunk`
2. Solved some auto-merge conflicts (by removing some un-needed commits after #50666 got merged)
3. Added one missing test case (to reach 100% coverage)

No need for an ETK release, as this PR only affect tests. Failing e2e test don't seem connected to changes in this PR and are happening on `trunk` as well.
